### PR TITLE
Module 4

### PR DESCRIPTION
**Pull request by Elder-HD**

### CODE REVIEW

File: src/main/java/org/example/CacheMain.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+public class CacheMain {
+
+
+    public static void main(String[] args) {
+        fullCacheTest();
+//        lfuCacheTest(); // run 2 times to see getting cached values
+    }
+
+    private static void lfuCacheTest() {
```
Comment (Volodymyr95) [inline src/main/java/org/example/CacheMain.java]: why all code in main?

File: src/main/java/org/example/cache/JedisLFUCache.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+        LOGGER.info("Connection with Jedis client established. {} DB is using", db);
+    }
+
+    public boolean isCityInCache(String cityKey) {
+        return jedis.exists(cityKey);
+    }
+
+    public City getCityFromCache(String cityKey) {
+        jedis.hincrBy(cityKey, "count", 1);
+        LOGGER.debug("City with {} was requested {} times", cityKey, jedis.hget(cityKey, "count"));
```
Comment (Volodymyr95) [inline src/main/java/org/example/cache/JedisLFUCache.java]: don't use debug, use info

File: src/main/java/org/example/cache/JedisLFUCache.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    private static final Logger LOGGER = LoggerFactory.getLogger(JedisLFUCache.class);
+    private final Jedis jedis;
+    private final CityMapper cityMapper;
+    private final ObjectMapper jsonMapper;
+    public final int CACHE_CAPACITY = 5;
+
+    public JedisLFUCache(int db) {
+        this.cityMapper = new CityMapper();
+        this.jsonMapper = new ObjectMapper();
+        this.jedis = new Jedis("localhost", 6379);
```
Comment (Volodymyr95) [inline src/main/java/org/example/cache/JedisLFUCache.java]: make port like constant

File: src/main/java/org/example/service/FullCacheService.java [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+        cityMapper = new CityMapper();
+    }
+
+
+    public List<City> fetchData() {
+        List<City> allCities = new ArrayList<>();
+        int totalCount = cityRepository.getTotalCount();
+        int step = 500;
+        for (int i = 0; i < totalCount; i += step) {
+            allCities.addAll(cityRepository.getItems(i, step));
```
Comment (Volodymyr95) [inline src/main/java/org/example/service/FullCacheService.java]: why u calling sql inside for? It's wrong, re-write using sql

Comment (Elder-HD) [reply src/main/java/org/example/service/FullCacheService.java]: From guide lecture: 
> Особливість реалізації така, що міста ми отримуємо по 500 штук. Це потрібно, бо існують обмеження на обсяг даних, що передаються. Так, в нашому випадку вони нас не стосуються, оскільки в нашій БД загалом 4079 міст. Але в продакшн застосунках, коли потрібно отримати багато даних, таким прийомом користуються часто.

But for this purpose one session and one transaction should be used (in my case all repository methods are transactional)
**Replaced with "getAll()" method.**

File: src/main/java/org/example/CacheMain.java [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+import org.example.service.CityService;
+import org.example.service.FullCacheService;
+
+public class CacheMain {
+
+
+    public static void main(String[] args) {
+        FullCacheService fullCache = new FullCacheService();
+        fullCache.fullCacheTest();
+//        lfuCacheTest(); // run 2 times to see getting cached values
```
Comment (Volodymyr95) [inline src/main/java/org/example/CacheMain.java]: don't push commented code

File: src/main/java/org/example/CacheMain.java [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+    }
+
+    private static void lfuCacheTest() {
+        CityService cityService = new CityService(new CityRepository(), new JedisLFUCache(1));
+        cityService.getCityById(1);
```
Comment (Volodymyr95) [inline src/main/java/org/example/CacheMain.java]: what mean all these numbers?

File: src/main/java/org/example/cache/JedisLFUCache.java [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+import redis.clients.jedis.Jedis;
+
+import java.util.Set;
+
+public class JedisLFUCache {
+    private static final Logger LOGGER = LoggerFactory.getLogger(JedisLFUCache.class);
+    private final Jedis jedis;
+    private final CityMapper cityMapper;
+    private final ObjectMapper jsonMapper;
+    private static final int PORT = 6379;
```
Comment (Volodymyr95) [inline src/main/java/org/example/cache/JedisLFUCache.java]: firs constants, then variables

File: src/main/java/org/example/cache/JedisLFUCache.java [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+        jedis.hincrBy(cityKey, "count", 1);
+        LOGGER.info("City with key \"{}\" was requested {} times", cityKey, jedis.hget(cityKey, "count"));
+        City city;
+        try {
+            String cityJson = jedis.hget(cityKey, "value");
+            CityCountry cityCountry = jsonMapper.readValue(cityJson, CityCountry.class);
+            city = cityMapper.toCity(cityCountry);
+        } catch (JsonProcessingException e) {
+            LOGGER.error("Error during Json deserialization for city {}", cityKey, e.getCause());
+            throw new RuntimeException("Error during Json deserialization");
```
Comment (Volodymyr95) [inline src/main/java/org/example/cache/JedisLFUCache.java]: create u own exception for this case

File: src/main/java/org/example/CacheMain.java (hunk 1 above)
Comment (Elder-HD) [reply src/main/java/org/example/CacheMain.java]: moved to FullCacheService class

File: src/main/java/org/example/cache/JedisLFUCache.java (hunk 2 above)
Comment (Elder-HD) [reply src/main/java/org/example/cache/JedisLFUCache.java]: Fixed

File: src/main/java/org/example/cache/JedisLFUCache.java (hunk 3 above)
Comment (Elder-HD) [reply src/main/java/org/example/cache/JedisLFUCache.java]: fixed

File: src/main/java/org/example/cache/JedisLFUCache.java (hunk 7 above)
Comment (Elder-HD) [reply src/main/java/org/example/cache/JedisLFUCache.java]: Fixed

File: src/main/java/org/example/cache/JedisLFUCache.java (hunk 8 above)
Comment (Elder-HD) [reply src/main/java/org/example/cache/JedisLFUCache.java]: changed to EntityConversionException

File: src/main/java/org/example/CacheMain.java (hunk 5 above)
Comment (Elder-HD) [reply src/main/java/org/example/CacheMain.java]: Uncommented

File: src/main/java/org/example/CacheMain.java (hunk 6 above)
Comment (Elder-HD) [reply src/main/java/org/example/CacheMain.java]: `JedisLFUCache` constructor with a parameter `int db` - invokes Jedis method `public String select(final int index)` after its initialization. 
Should the DB variable be defined as a constant?

`private static void lfuCacheTest()` method - just "demo" method using LFU caching strategy. 
Should I keep it without extracting it into a separate method?
<details><summary>Explanation</summary>
<p>
At the start, the application gets selected cities from the MySQL DB and caches it.
Each time the <code>getCityById(int id)</code> method is called, a check is made to retrieve the value from the cache.
If the city is already cached, it is returned as the result.
</p>
</details>

Comment (Elder-HD) [reply src/main/java/org/example/CacheMain.java]: Added `CacheService` interface with `process` method.
This method was implemented in both strategies.
All neccessery invokes were moved from `main` to those methods.

File: src/main/java/org/example/cache/CityCountry.java [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+import lombok.Getter;
+import lombok.Setter;
+import org.example.domain.enums.Continent;
+
+import java.math.BigDecimal;
+import java.util.Set;
+
+@Getter
+@Setter
+public class CityCountry {
```
Comment (Volodymyr95) [inline src/main/java/org/example/cache/CityCountry.java:12]: why not city?

**Review by Volodymyr95 — APPROVED** [review]

File: src/main/java/org/example/cache/CityCountry.java (hunk 9 above)
Comment (Elder-HD) [reply src/main/java/org/example/cache/CityCountry.java:12]: This class is used for caching and contains the fields of both entities (City and Country).

**Pull request merged**